crates/bitcoind_rpc/src/lib.rs

```diff
@@ -0,0 +1,476 @@
+//!
+//!     if update.is_block() {
+//!         let cp = update.checkpoint();
+//!         println!("block {}:{}", cp.height(), cp.hash());
+//!     } else {
+//!         println!("mempool!");
+//!     }
+//! }
+//! ```
+//!
+//! Alternatively, if the caller wishes to keep [`Emitter`] in a dedicated update-thread, the caller
+//! can continue to poll [`Iterator::next`] with a delay.
+
+#![warn(missing_docs)]
+
+use bdk_chain::{
+    bitcoin::{Block, Transaction},
+    indexed_tx_graph::Indexer,
+    local_chain::CheckPoint,
+    Append, BlockId, ConfirmationHeightAnchor, ConfirmationTimeAnchor, TxGraph,
+};
+pub use bitcoincore_rpc;
+use bitcoincore_rpc::{json::GetBlockResult, RpcApi};
+use std::fmt::Debug;
+
+/// An update emitted from [`Emitter`]. This can either be of a block or a subset of
+/// mempool transactions.
+#[derive(Debug, Clone)]
+pub enum EmittedUpdate {
+    /// An emitted block.
+    Block(EmittedBlock),
+    /// An emitted subset of mempool transactions.
+    ///
+    /// [`Emitter`] attempts to avoid re-emitting transactions.
+    Mempool(EmittedMempool),
+}
+
+impl EmittedUpdate {
+    /// Returns whether the update is of a subset of the mempool.
+    pub fn is_mempool(&self) -> bool {
+        matches!(self, Self::Mempool { .. })
+    }
+
+    /// Returns whether the update is of a block.
+    pub fn is_block(&self) -> bool {
+        matches!(self, Self::Block { .. })
+    }
+
+    /// Get the emission's checkpoint.
+    pub fn checkpoint(&self) -> CheckPoint {
+        match self {
+            EmittedUpdate::Block(e) => e.checkpoint(),
+            EmittedUpdate::Mempool(e) => e.checkpoint(),
+        }
+    }
+
+    /// Transforms the emitted update into a [`TxGraph`] update.
+    ///
+    /// The `tx_filter` parameter takes in a closure that filters out irrelevant transactions so
+    /// they do not get included in the [`TxGraph`] update. We have provided two closures;
+    /// [`empty_filter`] and [`indexer_filter`] for this purpose.
+    ///
+    /// The `anchor_map` parameter takes in a closure that creates anchors of a specific type.
+    /// [`confirmation_height_anchor`] and [`confirmation_time_anchor`] are avaliable to create
+    /// updates with [`ConfirmationHeightAnchor`] and [`ConfirmationTimeAnchor`] respectively.
+    pub fn into_tx_graph_update<F, M, A>(self, tx_filter: F, anchor_map: M) -> TxGraph<A>
+    where
+        F: FnMut(&Transaction) -> bool,
+        M: Fn(&CheckPoint, &Block, usize) -> A,
+        A: Clone + Ord + PartialOrd,
+    {
+        match self {
+            EmittedUpdate::Block(e) => e.into_tx_graph_update(tx_filter, anchor_map),
+            EmittedUpdate::Mempool(e) => e.into_tx_graph_update(tx_filter),
+        }
+    }
+}
+
+/// An emitted block.
+#[derive(Debug, Clone)]
+pub struct EmittedBlock {
+    /// The checkpoint constructed from the block's height/hash and connected to the previous block.
+    pub cp: CheckPoint,
```

Comment:
Is this needed. Won't the logic receiving this block be able to figure out where to connect it. `Block` has the previous block hash in it which should have already been emitted if the caller didn't know about it already.

Reply:
I think it makes our API a bit more flexible.

For example:
* The caller pulled data from esplora to height 100.
* The caller waits a while.
* The caller creates wallet with birthday at height 200.

If we create a `local_chain::Update` at height 200 with checkpoints derived from `Block` only, we can't connect to the chain at height 100.

Reply:
> I think it makes our API a bit more flexible.
> 
> For example:
> 
>     * The caller pulled data from esplora to height 100.
> 
>     * The caller waits a while.
> 
>     * The caller creates wallet with birthday at height 200.
> 
> 
> If we create a `local_chain::Update` at height 200 with checkpoints derived from `Block` only, we can't connect to the chain at height 100.

In this case you would pass in the block at `100` as `last_cp` to the emitter and then it would emit every block from 100 to 200 in order which means it would connect to `100`. Am I missing something?

Reply:
What if we want to sync from 200 without removing old blocks?

What if the caller only has the Genesis block and wallet birthday is at height 805016

Reply:
These are certainly questions. What are the answers? :sweat_smile:

Reply:
@LLFourn I'm very confused what you mean by this comment haha.

Reply:
I am trying to find out the answer to my question and the questions you posed in response to my question.